### PR TITLE
IRGen: Don't assert on keypaths that use a subclass as root

### PR DESCRIPTION
**Pull request by aschwaighofer**
class A {
   var x : Int = 0
 }

 class B : A {}

 _ = \B.x

rdar://46562046

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@jckarter Is that the right way to fix this?

**Comment by aschwaighofer** [comment]
getClassFieldAccess expects the class type that contains the field - it does not handle subclasses

File: lib/IRGen/GenKeyPath.cpp [hunk 1]
```diff
@@ -837,6 +837,21 @@ emitKeyPathComponent(IRGenModule &IGM,
     // ObjC-ness and resilience of the class hierarchy, there might be a few
     // different ways we need to go about this.
     if (loweredBaseTy.getClassOrBoundGenericClass()) {
+
+      // Walk up the class hierarchy to find the class decl that matches the
+      // property's class decl and use this type to determin the field access.
+      auto propertyBaseDecl = property->getDeclContext()->getSelfClassDecl();
+      auto currentBaseTy = loweredBaseTy.getASTType();
+      while (currentBaseTy->getClassOrBoundGenericClass() != propertyBaseDecl &&
```
Comment (slavapestov) [inline lib/IRGen/GenKeyPath.cpp]: This loop is already encapsulated in `TypeBase::getSuperclassForDecl()`

File: lib/IRGen/GenKeyPath.cpp [hunk 2]
```diff
@@ -837,6 +837,21 @@ emitKeyPathComponent(IRGenModule &IGM,
     // ObjC-ness and resilience of the class hierarchy, there might be a few
     // different ways we need to go about this.
     if (loweredBaseTy.getClassOrBoundGenericClass()) {
+
+      // Walk up the class hierarchy to find the class decl that matches the
+      // property's class decl and use this type to determin the field access.
```
[the remark this answers is not in the document]
Comment (jckarter) [reply lib/IRGen/GenKeyPath.cpp]: That might be cleaner, yeah. Then we can assert the condition in the SILVerifier too.

**Comment by jckarter** [comment]
This SGTM, though I agree with Slava that having SIL always reference the property by its originating superclass is a little bit cleaner.

**Comment by aschwaighofer** [comment]
Consider the following example:
```
$ cat KP3.swift
class A {
let x : Int = 0
}

class B : A {}

class C {
  let b: B = B()
}

func test() {
 _ = \C.b.x
}
```
Today the key path is
```
keypath $KeyPath<C, Int>, (root $C; stored_property #C.b : $B; stored_property #A.x : $Int)
```
If we change SILGen this would be:
```
keypath $KeyPath<C, Int>, (root $C; stored_property #C.b : $A; stored_property #A.x : $Int)
```
Which would seem odd to me that the type for the var decl:`#C.b` disagrees with the type we assign the component: `$A`.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
Ah, I misunderstood, I thought that the `stored_property` key had been ending up as `#B.x`. In that case, I think your fix is correct.

**Comment by slavapestov** [comment]
@aschwaighofer Thanks for the cleanup! If this same pattern comes up in more places, we could add a SILType::getSuperclassDecl() variant too, but it's not too bad as-is.

**Pull request merged**